lib/compiler-cranelift/src/translator/code_translator.rs

```diff
@@ -78,27 +78,35 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
          ***********************************************************************************/
         Operator::LocalGet { local_index } => {
             let val = builder.use_var(Variable::with_u32(*local_index));
-            state.push1(val);
+            let local_type = environ.get_local_type(*local_index).unwrap();
+            let ref_counted = local_type == WasmerType::ExternRef;
+            state.push1_extra((
+                val,
+                ValueExtraInfo {
+                    ref_counted,
+                    ..Default::default()
+                },
+            ));
             let label = ValueLabel::from_u32(*local_index);
             builder.set_val_label(val, label);
 
-            let local_type = environ.get_local_type(*local_index).unwrap();
-            if local_type == WasmerType::ExternRef {
+            if ref_counted {
                 environ.translate_externref_inc(builder.cursor(), val)?;
             }
         }
         Operator::LocalSet { local_index } => {
-            let mut val = state.pop1();
+            let (mut val, _metadata) = state.pop1();
 
             // Ensure SIMD values are cast to their default Cranelift type, I8x16.
             let ty = builder.func.dfg.value_type(val);
             let local_type = environ.get_local_type(*local_index).unwrap();
             if local_type == WasmerType::ExternRef {
-                environ.translate_externref_dec(builder.cursor(), val)?;
-            }
-            if ty.is_ref() {
+                debug_assert!(_metadata.ref_counted);
+                let existing_val = builder.use_var(Variable::with_u32(*local_index));
+                environ.translate_externref_dec(builder.cursor(), existing_val)?;
                 environ.translate_externref_inc(builder.cursor(), val)?;
```

Comment:
The value already has one use on the local value stack, and then local.set drops it (dec-ref) and it lives on in the local value slot (inc-ref). That dec+inc is a no-op so we shouldn't be inc-ref'ing it here.

Also, always inc before dec please. That's not the bug here, but in general it prevents problems when val == existing_val.

Reply:
Good call on the `inc` and ordering!